crates/lsp/src/lib.rs

```diff
@@ -446,158 +490,61 @@ impl<L: LSPLang> Backend<L> {
-  use ast_grep_config::{from_yaml_string, GlobalRules};
-  use ast_grep_language::SupportLang;
-  use serde_json::Value;
-  use tokio::io::{duplex, AsyncReadExt, AsyncWriteExt, DuplexStream};
-
-  fn start_lsp() -> (DuplexStream, DuplexStream) {
-    let globals = GlobalRules::default();
-    let config: RuleConfig<SupportLang> = from_yaml_string(
-      r"
-id: no-console-rule
-message: No console.log
-severity: warning
-language: TypeScript
-rule:
-  pattern: console.log($$$A)
-note: no console.log
-fix: |
-  alert($$$A)
-",
-      &globals,
-    )
-    .unwrap()
-    .pop()
-    .unwrap();
-    let rc: RuleCollection<SupportLang> = RuleCollection::try_new(vec![config]).unwrap();
-    let rc_result: std::result::Result<_, String> = Ok(rc);
-    let (service, socket) = LspService::build(|client| Backend::new(client, rc_result)).finish();
-    let (req_client, req_server) = duplex(1024);
-    let (resp_server, resp_client) = duplex(1024);
-
-    // start server as concurrent task
-    tokio::spawn(Server::new(req_server, resp_server, socket).serve(service));
-
-    (req_client, resp_client)
-  }
-
-  fn req(msg: &str) -> String {
-    format!("Content-Length: {}\r\n\r\n{}", msg.len(), msg)
-  }
-
-  // A function that takes a byte slice as input and returns the content length as an option
-  fn resp(input: &[u8]) -> Option<&str> {
-    let input_str = std::str::from_utf8(input).ok()?;
-    let mut splits = input_str.split("\r\n\r\n");
-    let header = splits.next()?;
-    let body = splits.next()?;
-    let length_str = header.trim_start_matches("Content-Length: ");
-    let length = length_str.parse::<usize>().ok()?;
-    Some(&body[..length])
-  }
-
-  async fn test_lsp() {
-    let initialize = r#"{
-      "jsonrpc":"2.0",
-      "id": 1,
-      "method": "initialize",
-      "params": {
-        "capabilities": {
-          "textDocumentSync": 1
-        }
-      }
-    }"#;
-    let (mut req_client, mut resp_client) = start_lsp();
-    let mut buf = vec![0; 1024];
 
-    req_client
-      .write_all(req(initialize).as_bytes())
-      .await
-      .unwrap();
-    let _ = resp_client.read(&mut buf).await.unwrap();
-
-    assert!(resp(&buf).unwrap().starts_with('{'));
-
-    let save_file = r#"{
-  "jsonrpc": "2.0",
-  "id": 1,
-  "method": "textDocument/codeAction",
-  "params": {
-    "range": {
-      "end": {
-        "character": 10,
-        "line": 1
-      },
-      "start": {
-        "character": 10,
-        "line": 1
+  async fn on_execute_command(&self, params: ExecuteCommandParams) -> Option<Value> {
+    let ExecuteCommandParams {
+      arguments,
+      command,
+      work_done_progress_params: _,
+    } = params;
+
+    match command.as_ref() {
+      APPLY_ALL_FIXES => {
+        self
+          .client
+          .log_message(
+            MessageType::LOG,
+            format!("Running ExecuteCommand {}", command),
+          )
+          .await;
+        let uri = Url::parse(arguments.first()?["uri"].as_str()?).ok()?;
```

Comment:
This code looks pretty weird. We should add a serde struct to decode the JSON to a Rust struct?